dagstore/dagstorewrapper.go

```diff
@@ -0,0 +1,101 @@
+package dagstore
+
+import (
+	"context"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/dagstore"
+	"github.com/filecoin-project/dagstore/mount"
+	"github.com/filecoin-project/dagstore/shard"
+
+	"github.com/filecoin-project/go-fil-markets/carstore"
+)
+
+// DagStoreWrapper hides the details of the DAG store implementation from
+// the other parts of go-fil-markets
+type DagStoreWrapper interface {
+	// RegisterShard loads a CAR file into the DAG store and builds an index for it
+	RegisterShard(ctx context.Context, pieceCid cid.Cid, carPath string) error
+	// LoadShard fetches the data for a shard and provides a blockstore interface to it
+	LoadShard(ctx context.Context, pieceCid cid.Cid) (carstore.ClosableBlockstore, error)
+}
+
+type dagStoreWrapper struct {
+	dagStore *dagstore.DAGStore
+	mountApi LotusMountAPI
+}
+
+func NewDagStoreWrapper(dsRegistry *mount.Registry, dagStore *dagstore.DAGStore, mountApi LotusMountAPI) (*dagStoreWrapper, error) {
+	err := dsRegistry.Register(lotusScheme, NewLotusMountTemplate(mountApi))
+	if err != nil {
+		return nil, err
+	}
+
+	return &dagStoreWrapper{
+		dagStore: dagStore,
+		mountApi: mountApi,
+	}, nil
+}
+
+type closableBlockstore struct {
+	dagstore.ReadBlockstore
+	io.Closer
+}
+
+func (ds *dagStoreWrapper) LoadShard(ctx context.Context, pieceCid cid.Cid) (carstore.ClosableBlockstore, error) {
+	key := shard.KeyFromCID(pieceCid)
+	resch := make(chan dagstore.ShardResult, 1)
+	err := ds.dagStore.AcquireShard(ctx, key, resch, dagstore.AcquireOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Can I rely on AcquireShard to return an error if the context times out?
+	//select {
+	//case <-ctx.Done():
+	//	return ctx.Err()
+	//case res := <-resch:
+	//	return nil, res.Error
+	//}
```

Comment:
Not yet. If you need this feature we need to prioritise this: https://github.com/filecoin-project/dagstore/issues/39